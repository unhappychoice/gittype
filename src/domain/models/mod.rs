pub mod challenge;
pub mod chunk;
pub mod color_mode;
pub mod color_scheme;
pub mod config;
pub mod countdown;
pub mod extraction_options;
pub mod git_repository;
pub mod git_repository_ref;
pub mod language;
pub mod languages;
pub mod rank;
pub mod session;
pub mod stage;
pub mod storage;
pub mod theme;
pub mod total;
pub mod version;

// Re-export main types for easy access
pub use challenge::Challenge;
pub use chunk::{ChunkType, CodeChunk};
pub use countdown::Countdown;
pub use extraction_options::ExtractionOptions;
pub use git_repository::GitRepository;
pub use git_repository_ref::GitRepositoryRef;
pub use language::{Language, Languages};
pub use rank::{Rank, RankTier};
pub use session::{Session, SessionResult};
pub use stage::{Stage, StageResult};
pub use total::{Total, TotalResult};
