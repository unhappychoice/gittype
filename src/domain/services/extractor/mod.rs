pub mod challenge_converter;
pub mod code_chunk_extractor;
pub mod core;
pub mod git_repository_extractor;
pub mod language_registry;
pub mod parsers;
pub mod repository_extractor;

pub use challenge_converter::ChallengeConverter;
pub use code_chunk_extractor::CodeChunkExtractor;
pub use git_repository_extractor::GitRepositoryExtractor;
pub use language_registry::LanguageRegistry;
pub use repository_extractor::RepositoryExtractor;
