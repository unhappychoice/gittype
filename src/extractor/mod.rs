pub mod challenge_converter;
pub mod code_chunk_extractor;
pub mod core;
pub mod git_repository_extractor;
pub mod models;
pub mod parsers;
pub mod repository_extractor;

pub use crate::game::screens::loading_screen::{NoOpProgressReporter, ProgressReporter};
pub use crate::models::{ChunkType, CodeChunk, GitRepository};
pub use challenge_converter::ChallengeConverter;
pub use code_chunk_extractor::CodeChunkExtractor;
pub use git_repository_extractor::GitRepositoryExtractor;
pub use models::{ExtractionOptions, Language};
pub use repository_extractor::RepositoryExtractor;
