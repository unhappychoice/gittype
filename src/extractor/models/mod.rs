pub mod chunk;
pub mod language;
pub mod options;

pub use chunk::{ChunkType, CodeChunk};
pub use language::Language;
pub use options::ExtractionOptions;
