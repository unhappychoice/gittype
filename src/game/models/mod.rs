pub mod loading_steps;

pub use loading_steps::*;
