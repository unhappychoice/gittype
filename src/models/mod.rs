pub mod chunk;
pub mod challenge;
pub mod stage;
pub mod session;
pub mod total;
pub mod rank;
pub mod git_repository;

// Re-export main types for easy access
pub use chunk::{ChunkType, CodeChunk};
pub use challenge::Challenge;
pub use stage::{Stage, StageResult};
pub use session::{Session, SessionResult};
pub use total::{Total, TotalResult};
pub use rank::{Rank, RankingTitle};
pub use git_repository::GitRepository;