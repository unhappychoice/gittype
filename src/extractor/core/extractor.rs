use crate::extractor::parsers::{get_parser_registry, parse_with_thread_local};
use crate::models::CodeChunk;
use crate::{GitTypeError, Result};
use std::fs;
use std::path::Path;
use streaming_iterator::StreamingIterator;
use tree_sitter::{Node, QueryCursor, Tree};

pub struct CommonExtractor;

impl CommonExtractor {
    pub fn extract_chunks_from_tree(
        tree: &Tree,
        source_code: &str,
        file_path: &Path,
        language: &str,
    ) -> Result<Vec<CodeChunk>> {
        let mut chunks = Vec::new();
        let registry = get_parser_registry();

        let file_comment_ranges = Self::extract_comment_ranges(tree, source_code, language)?;
        let query = registry.create_query(language)?;

        let mut cursor = QueryCursor::new();
        let mut matches = cursor.matches(&query, tree.root_node(), source_code.as_bytes());

        while let Some(match_) = matches.next() {
            for capture in match_.captures {
                let node = capture.node;
                let capture_name = &query.capture_names()[capture.index as usize];

                if let Some(chunk) = Self::node_to_chunk(
                    node,
                    source_code,
                    file_path,
                    language,
                    capture_name,
                    &file_comment_ranges,
                ) {
                    chunks.push(chunk);
                }
            }
        }

        Ok(chunks)
    }

    pub fn extract_from_file(file_path: &Path, language: &str) -> Result<Vec<CodeChunk>> {
        let content = fs::read_to_string(file_path)?;
        // Reuse per-thread parser instance for the language
        let tree = parse_with_thread_local(language, &content).ok_or_else(|| {
            GitTypeError::ExtractionFailed(format!("Failed to parse file: {:?}", file_path))
        })?;

        Self::extract_chunks_from_tree(&tree, &content, file_path, language)
    }

    pub fn extract_comment_ranges(
        tree: &Tree,
        source_code: &str,
        language: &str,
    ) -> Result<Vec<(usize, usize)>> {
        let registry = get_parser_registry();
        let comment_query = registry.create_comment_query(language)?;
        let mut comment_ranges = Vec::new();

        let mut cursor = QueryCursor::new();
        let mut matches = cursor.matches(&comment_query, tree.root_node(), source_code.as_bytes());

        while let Some(m) = matches.next() {
            for capture in m.captures {
                let node = capture.node;
                let start = node.start_byte();
                let end = node.end_byte();

                if Self::is_valid_comment_node(node, language) {
                    comment_ranges.push((start, end));
                }
            }
        }

        comment_ranges.sort_by_key(|&(start, _)| start);
        Ok(comment_ranges)
    }

    fn is_valid_comment_node(node: Node, language: &str) -> bool {
        let node_kind = node.kind();
        match language {
            "rust" => node_kind == "line_comment" || node_kind == "block_comment",
            "typescript" => node_kind == "comment",
            "javascript" => node_kind == "comment",
            "python" => node_kind == "comment",
            "ruby" => node_kind == "comment",
            "go" => node_kind == "comment",
            "swift" => node_kind == "comment" || node_kind == "multiline_comment",
            "kotlin" => node_kind == "line_comment" || node_kind == "multiline_comment",
            "java" => node_kind == "line_comment" || node_kind == "block_comment",
            "php" => node_kind == "comment" || node_kind == "shell_comment_line",
            "csharp" => node_kind == "comment",
            "c" => node_kind == "comment",
            "cpp" => node_kind == "comment",
            "haskell" => node_kind == "comment",
            "dart" => node_kind == "comment" || node_kind == "documentation_comment",
            _ => false,
        }
    }

    fn node_to_chunk(
        node: Node,
        source_code: &str,
        file_path: &Path,
        language: &str,
        capture_name: &str,
        file_comment_ranges: &[(usize, usize)],
    ) -> Option<CodeChunk> {
        let start_byte = node.start_byte();
        let end_byte = node.end_byte();
        let content = &source_code[start_byte..end_byte];

        let start_line = node.start_position().row + 1;
        let end_line = node.end_position().row + 1;
        let original_indentation = node.start_position().column;

        // Extract actual indentation characters from source
        let original_indent_chars = if original_indentation > 0 {
            Self::extract_line_indent_chars(
                source_code,
                node.start_position().row,
                original_indentation,
            )
        } else {
            String::new()
        };

        let registry = get_parser_registry();
        let extractor = registry.get_extractor(language).ok()?;

        let chunk_type = extractor.capture_name_to_chunk_type(capture_name)?;

        let name = extractor
            .extract_name(node, source_code, capture_name)
            .or_else(|| Self::extract_name(node, source_code))
            .unwrap_or_else(|| "unknown".to_string());

        let chunk_comment_ranges: Vec<(usize, usize)> = file_comment_ranges
            .iter()
            .filter_map(|&(comment_start, comment_end)| {
                if comment_start >= start_byte && comment_end <= end_byte {
                    Some((comment_start - start_byte, comment_end - start_byte))
                } else {
                    None
                }
            })
            .collect();

        let normalized_content = Self::normalize_first_line_indentation(
            content,
            &original_indent_chars,
        );

        Some(CodeChunk {
            content: normalized_content,
            file_path: file_path.to_path_buf(),
            start_line,
            end_line,
            language: language.to_string(),
            chunk_type,
            name,
            comment_ranges: chunk_comment_ranges,
            original_indentation,
        })
    }

    fn extract_name(node: Node, source_code: &str) -> Option<String> {
        if node.kind() == "variable_declarator" {
            let mut cursor = node.walk();
            if cursor.goto_first_child() {
                let name_node = cursor.node();
                if name_node.kind() == "identifier" {
                    let start = name_node.start_byte();
                    let end = name_node.end_byte();
                    return Some(source_code[start..end].to_string());
                }
            }
        }

        let mut cursor = node.walk();
        if cursor.goto_first_child() {
            loop {
                let child = cursor.node();
                if child.kind() == "identifier"
                    || child.kind() == "type_identifier"
                    || child.kind() == "property_identifier"
                    || child.kind() == "field_identifier"
                    || child.kind() == "constant"
                    || child.kind() == "simple_identifier"
                {
                    let start = child.start_byte();
                    let end = child.end_byte();
                    return Some(source_code[start..end].to_string());
                }
                if !cursor.goto_next_sibling() {
                    break;
                }
            }
        }
        None
    }

    fn normalize_first_line_indentation(
        content: &str,
        original_indent_chars: &str,
    ) -> String {
        let lines: Vec<&str> = content.lines().collect();
        if lines.is_empty() {
            return content.to_string();
        }

        let mut result_lines = Vec::new();
        
        for (line_idx, line) in lines.iter().enumerate() {
            if line_idx == 0 {
                // First line: add original indentation characters from source
                result_lines.push(format!("{}{}", original_indent_chars, line));
            } else {
                // Other lines: keep as is
                result_lines.push(line.to_string());
            }
        }

        result_lines.join("\n")
    }

    fn extract_line_indent_chars(
        source_code: &str,
        line_row: usize,
        indent_length: usize,
    ) -> String {
        let lines: Vec<&str> = source_code.lines().collect();
        if line_row < lines.len() {
            let line = lines[line_row];
            line.chars().take(indent_length).collect()
        } else {
            String::new()
        }
    }
}
