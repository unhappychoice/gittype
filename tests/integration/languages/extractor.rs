// Don't import here since they're used inside the macro
// use these inside the macro expansion

/// Test macro for language extractors with snapshot functionality
macro_rules! test_language_extractor {
    (
        name: $test_name:ident,
        language: $language:literal,
        extension: $extension:literal,
        source: $source:expr,
        total_chunks: $total:expr,
        chunk_counts: {
            $($chunk_type:ident: $count:expr),* $(,)?
        }
    ) => {
        #[test]
        fn $test_name() {
            use crate::integration::{extract_chunks_for_test, test_extraction_options};
            use gittype::extractor::CodeChunkExtractor;
            use gittype::models::ChunkType;
            use std::collections::HashMap;
            use std::fs;
            use tempfile::TempDir;

            let temp_dir = TempDir::new().unwrap();
            let file_path = temp_dir.path().join(concat!("test.", $extension));

            fs::write(&file_path, $source).unwrap();

            let mut extractor = CodeChunkExtractor::new().unwrap();
            let chunks = extract_chunks_for_test(
                &mut extractor,
                temp_dir.path(),
                test_extraction_options()
            ).unwrap();

            // Check total chunk count
            assert_eq!(
                chunks.len(),
                $total,
                "Expected {} total chunks, got {}",
                $total,
                chunks.len()
            );

            // Check chunk counts by type
            let mut chunk_counts: HashMap<ChunkType, usize> = HashMap::new();
            for chunk in &chunks {
                *chunk_counts.entry(chunk.chunk_type.clone()).or_insert(0) += 1;
            }

            $(
                let expected_count = $count;
                let actual_count = chunk_counts.get(&ChunkType::$chunk_type).copied().unwrap_or(0);
                assert_eq!(
                    actual_count,
                    expected_count,
                    "Expected {} {:?} chunks, got {}",
                    expected_count,
                    ChunkType::$chunk_type,
                    actual_count
                );
            )*

            // Verify all chunks have correct language
            for chunk in &chunks {
                assert_eq!(chunk.language, $language.to_string());
            }

            // Always create snapshot
            crate::integration::languages::extractor::create_insta_snapshot(&chunks, $source, stringify!($test_name));
        }
    };
}

/// Create a snapshot using insta
pub fn create_insta_snapshot(
    chunks: &[gittype::models::CodeChunk],
    source_code: &str,
    test_name: &str,
) {
    use insta::assert_snapshot;
    use serde_json::json;

    let snapshot_data = json!({
        "test_name": test_name,
        "source_code": source_code,
        "total_chunks": chunks.len(),
        "chunks": chunks.iter().map(|chunk| {
            json!({
                "chunk_type": format!("{:?}", chunk.chunk_type),
                "name": chunk.name,
                "language": chunk.language,
                "start_line": chunk.start_line,
                "end_line": chunk.end_line,
                "original_indentation": chunk.original_indentation,
                "comment_ranges": chunk.comment_ranges,
                "content": chunk.content
            })
        }).collect::<Vec<_>>()
    });

    assert_snapshot!(
        test_name,
        serde_json::to_string_pretty(&snapshot_data).unwrap()
    );
}

// Export the macro for use in other modules
pub(crate) use test_language_extractor;
