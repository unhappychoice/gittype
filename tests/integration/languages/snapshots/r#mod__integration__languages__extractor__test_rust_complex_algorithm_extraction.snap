---
source: tests/integration/languages/extractor.rs
expression: "serde_json::to_string_pretty(&snapshot_data).unwrap()"
---
{
  "chunks": [
    {
      "chunk_type": "File",
      "comment_ranges": [
        [
          185,
          235
        ],
        [
          496,
          536
        ],
        [
          866,
          887
        ],
        [
          1237,
          1293
        ],
        [
          1841,
          1866
        ]
      ],
      "content": "\nuse std::collections::HashMap;\n\nfn complex_data_processor(input: Vec<i32>, threshold: i32) -> HashMap<String, i32> {\n    let mut result = HashMap::new();\n    let mut counter = 0;\n\n    // Main processing loop - extractable middle chunk\n    for (index, value) in input.iter().enumerate() {\n        let key = format!(\"item_{}\", index);\n\n        if *value > threshold {\n            let processed = value * 2;\n            result.insert(key.clone(), processed);\n            counter += 1;\n\n            // Additional processing for high values\n            if processed > 100 {\n                let bonus_key = format!(\"{}_bonus\", key);\n                result.insert(bonus_key, processed / 10);\n            }\n        } else if *value > 0 {\n            let adjusted = value + threshold;\n            result.insert(key, adjusted);\n            counter += 1;\n        }\n    }\n\n    // Finalization logic\n    if counter > 0 {\n        result.insert(\"total_processed\".to_string(), counter);\n        result.insert(\"average_processed\".to_string(),\n                     result.values().sum::<i32>() / counter);\n    }\n\n    result\n}\n\npub fn advanced_string_matcher(patterns: &[&str], text: &str) -> Vec<(usize, String)> {\n    let mut matches = Vec::new();\n\n    // Pattern matching algorithm - extractable middle chunk\n    for (pattern_idx, pattern) in patterns.iter().enumerate() {\n        let mut search_start = 0;\n\n        while let Some(pos) = text[search_start..].find(pattern) {\n            let absolute_pos = search_start + pos;\n            let context_start = absolute_pos.saturating_sub(10);\n            let context_end = (absolute_pos + pattern.len() + 10).min(text.len());\n            let context = text[context_start..context_end].to_string();\n\n            matches.push((absolute_pos, context));\n            search_start = absolute_pos + 1;\n\n            // Prevent infinite loops\n            if search_start >= text.len() {\n                break;\n            }\n        }\n    }\n\n    matches.sort_by_key(|&(pos, _)| pos);\n    matches.dedup_by_key(|&mut (pos, _)| pos);\n    matches\n}\n",
      "end_line": 65,
      "language": "rust",
      "name": "entire_file",
      "original_indentation": 0,
      "start_line": 1
    },
    {
      "chunk_type": "CodeBlock",
      "comment_ranges": [],
      "content": "fn complex_data_processor",
      "end_line": 4,
      "language": "rust",
      "name": "name",
      "original_indentation": 3,
      "start_line": 4
    },
    {
      "chunk_type": "Function",
      "comment_ranges": [
        [
          152,
          202
        ],
        [
          463,
          503
        ],
        [
          833,
          854
        ]
      ],
      "content": "fn complex_data_processor(input: Vec<i32>, threshold: i32) -> HashMap<String, i32> {\n    let mut result = HashMap::new();\n    let mut counter = 0;\n\n    // Main processing loop - extractable middle chunk\n    for (index, value) in input.iter().enumerate() {\n        let key = format!(\"item_{}\", index);\n\n        if *value > threshold {\n            let processed = value * 2;\n            result.insert(key.clone(), processed);\n            counter += 1;\n\n            // Additional processing for high values\n            if processed > 100 {\n                let bonus_key = format!(\"{}_bonus\", key);\n                result.insert(bonus_key, processed / 10);\n            }\n        } else if *value > 0 {\n            let adjusted = value + threshold;\n            result.insert(key, adjusted);\n            counter += 1;\n        }\n    }\n\n    // Finalization logic\n    if counter > 0 {\n        result.insert(\"total_processed\".to_string(), counter);\n        result.insert(\"average_processed\".to_string(),\n                     result.values().sum::<i32>() / counter);\n    }\n\n    result\n}",
      "end_line": 37,
      "language": "rust",
      "name": "function",
      "original_indentation": 0,
      "start_line": 4
    },
    {
      "chunk_type": "Loop",
      "comment_ranges": [],
      "content": "    for (index, value) in input.iter().enumerate() {\n        let key = format!(\"item_{}\", index);\n\n        if *value > threshold {\n            let processed = value * 2;\n            result.insert(key.clone(), processed);\n            counter += 1;\n\n            // Additional processing for high values\n            if processed > 100 {\n                let bonus_key = format!(\"{}_bonus\", key);\n                result.insert(bonus_key, processed / 10);\n            }\n        } else if *value > 0 {\n            let adjusted = value + threshold;\n            result.insert(key, adjusted);\n            counter += 1;\n        }\n    }",
      "end_line": 27,
      "language": "rust",
      "name": "for_loop",
      "original_indentation": 4,
      "start_line": 9
    },
    {
      "chunk_type": "CodeBlock",
      "comment_ranges": [],
      "content": "        if *value > threshold {\n            let processed = value * 2;\n            result.insert(key.clone(), processed);\n            counter += 1;\n\n            // Additional processing for high values\n            if processed > 100 {\n                let bonus_key = format!(\"{}_bonus\", key);\n                result.insert(bonus_key, processed / 10);\n            }\n        }",
      "end_line": 22,
      "language": "rust",
      "name": "code_block",
      "original_indentation": 30,
      "start_line": 12
    },
    {
      "chunk_type": "Conditional",
      "comment_ranges": [],
      "content": "        if *value > threshold {\n            let processed = value * 2;\n            result.insert(key.clone(), processed);\n            counter += 1;\n\n            // Additional processing for high values\n            if processed > 100 {\n                let bonus_key = format!(\"{}_bonus\", key);\n                result.insert(bonus_key, processed / 10);\n            }\n        } else if *value > 0 {\n            let adjusted = value + threshold;\n            result.insert(key, adjusted);\n            counter += 1;\n        }",
      "end_line": 26,
      "language": "rust",
      "name": "if_block",
      "original_indentation": 8,
      "start_line": 12
    },
    {
      "chunk_type": "Conditional",
      "comment_ranges": [],
      "content": "            if processed > 100 {\n                let bonus_key = format!(\"{}_bonus\", key);\n                result.insert(bonus_key, processed / 10);\n            }",
      "end_line": 21,
      "language": "rust",
      "name": "if_block",
      "original_indentation": 12,
      "start_line": 18
    },
    {
      "chunk_type": "Conditional",
      "comment_ranges": [],
      "content": "        } else if *value > 0 {\n            let adjusted = value + threshold;\n            result.insert(key, adjusted);\n            counter += 1;\n        }",
      "end_line": 26,
      "language": "rust",
      "name": "if_block",
      "original_indentation": 15,
      "start_line": 22
    },
    {
      "chunk_type": "Conditional",
      "comment_ranges": [],
      "content": "    if counter > 0 {\n        result.insert(\"total_processed\".to_string(), counter);\n        result.insert(\"average_processed\".to_string(),\n                     result.values().sum::<i32>() / counter);\n    }",
      "end_line": 34,
      "language": "rust",
      "name": "if_block",
      "original_indentation": 4,
      "start_line": 30
    },
    {
      "chunk_type": "FunctionCall",
      "comment_ranges": [],
      "content": "        result.insert(\"average_processed\".to_string(),\n                     result.values().sum::<i32>() / counter)",
      "end_line": 33,
      "language": "rust",
      "name": "function_call",
      "original_indentation": 8,
      "start_line": 32
    },
    {
      "chunk_type": "CodeBlock",
      "comment_ranges": [],
      "content": "pub fn advanced_string_matcher",
      "end_line": 39,
      "language": "rust",
      "name": "name",
      "original_indentation": 7,
      "start_line": 39
    },
    {
      "chunk_type": "Function",
      "comment_ranges": [
        [
          127,
          183
        ],
        [
          731,
          756
        ]
      ],
      "content": "pub fn advanced_string_matcher(patterns: &[&str], text: &str) -> Vec<(usize, String)> {\n    let mut matches = Vec::new();\n\n    // Pattern matching algorithm - extractable middle chunk\n    for (pattern_idx, pattern) in patterns.iter().enumerate() {\n        let mut search_start = 0;\n\n        while let Some(pos) = text[search_start..].find(pattern) {\n            let absolute_pos = search_start + pos;\n            let context_start = absolute_pos.saturating_sub(10);\n            let context_end = (absolute_pos + pattern.len() + 10).min(text.len());\n            let context = text[context_start..context_end].to_string();\n\n            matches.push((absolute_pos, context));\n            search_start = absolute_pos + 1;\n\n            // Prevent infinite loops\n            if search_start >= text.len() {\n                break;\n            }\n        }\n    }\n\n    matches.sort_by_key(|&(pos, _)| pos);\n    matches.dedup_by_key(|&mut (pos, _)| pos);\n    matches\n}",
      "end_line": 65,
      "language": "rust",
      "name": "function",
      "original_indentation": 0,
      "start_line": 39
    },
    {
      "chunk_type": "Loop",
      "comment_ranges": [],
      "content": "    for (pattern_idx, pattern) in patterns.iter().enumerate() {\n        let mut search_start = 0;\n\n        while let Some(pos) = text[search_start..].find(pattern) {\n            let absolute_pos = search_start + pos;\n            let context_start = absolute_pos.saturating_sub(10);\n            let context_end = (absolute_pos + pattern.len() + 10).min(text.len());\n            let context = text[context_start..context_end].to_string();\n\n            matches.push((absolute_pos, context));\n            search_start = absolute_pos + 1;\n\n            // Prevent infinite loops\n            if search_start >= text.len() {\n                break;\n            }\n        }\n    }",
      "end_line": 60,
      "language": "rust",
      "name": "for_loop",
      "original_indentation": 4,
      "start_line": 43
    },
    {
      "chunk_type": "Loop",
      "comment_ranges": [],
      "content": "        while let Some(pos) = text[search_start..].find(pattern) {\n            let absolute_pos = search_start + pos;\n            let context_start = absolute_pos.saturating_sub(10);\n            let context_end = (absolute_pos + pattern.len() + 10).min(text.len());\n            let context = text[context_start..context_end].to_string();\n\n            matches.push((absolute_pos, context));\n            search_start = absolute_pos + 1;\n\n            // Prevent infinite loops\n            if search_start >= text.len() {\n                break;\n            }\n        }",
      "end_line": 59,
      "language": "rust",
      "name": "while_loop",
      "original_indentation": 8,
      "start_line": 46
    },
    {
      "chunk_type": "Conditional",
      "comment_ranges": [],
      "content": "            if search_start >= text.len() {\n                break;\n            }",
      "end_line": 58,
      "language": "rust",
      "name": "if_block",
      "original_indentation": 12,
      "start_line": 56
    }
  ],
  "source_code": "\nuse std::collections::HashMap;\n\nfn complex_data_processor(input: Vec<i32>, threshold: i32) -> HashMap<String, i32> {\n    let mut result = HashMap::new();\n    let mut counter = 0;\n\n    // Main processing loop - extractable middle chunk\n    for (index, value) in input.iter().enumerate() {\n        let key = format!(\"item_{}\", index);\n\n        if *value > threshold {\n            let processed = value * 2;\n            result.insert(key.clone(), processed);\n            counter += 1;\n\n            // Additional processing for high values\n            if processed > 100 {\n                let bonus_key = format!(\"{}_bonus\", key);\n                result.insert(bonus_key, processed / 10);\n            }\n        } else if *value > 0 {\n            let adjusted = value + threshold;\n            result.insert(key, adjusted);\n            counter += 1;\n        }\n    }\n\n    // Finalization logic\n    if counter > 0 {\n        result.insert(\"total_processed\".to_string(), counter);\n        result.insert(\"average_processed\".to_string(),\n                     result.values().sum::<i32>() / counter);\n    }\n\n    result\n}\n\npub fn advanced_string_matcher(patterns: &[&str], text: &str) -> Vec<(usize, String)> {\n    let mut matches = Vec::new();\n\n    // Pattern matching algorithm - extractable middle chunk\n    for (pattern_idx, pattern) in patterns.iter().enumerate() {\n        let mut search_start = 0;\n\n        while let Some(pos) = text[search_start..].find(pattern) {\n            let absolute_pos = search_start + pos;\n            let context_start = absolute_pos.saturating_sub(10);\n            let context_end = (absolute_pos + pattern.len() + 10).min(text.len());\n            let context = text[context_start..context_end].to_string();\n\n            matches.push((absolute_pos, context));\n            search_start = absolute_pos + 1;\n\n            // Prevent infinite loops\n            if search_start >= text.len() {\n                break;\n            }\n        }\n    }\n\n    matches.sort_by_key(|&(pos, _)| pos);\n    matches.dedup_by_key(|&mut (pos, _)| pos);\n    matches\n}\n",
  "test_name": "test_rust_complex_algorithm_extraction",
  "total_chunks": 15
}
